bot/exts/evergreen/duck_game.py

```diff
@@ -0,0 +1,329 @@
+import asyncio
+import random
+import re
+from collections import defaultdict
+from io import BytesIO
+from itertools import product
+from pathlib import Path
+from urllib.parse import urlparse
+
+import discord
+from PIL import Image, ImageDraw, ImageFont
+from discord.ext import commands
+
+from bot.bot import Bot
+from bot.constants import Colours, MODERATION_ROLES
+
+
+DECK = list(product(*[(0, 1, 2)]*4))
+
+GAME_DURATION = 180
+
+# Scoring
+CORRECT_SOLN = 1
+INCORRECT_SOLN = -1
+CORRECT_GOOSE = 2
+INCORRECT_GOOSE = -1
+
+""" Distribution of minimum acceptable solutions at board generation.
+    This is for gameplay reasons, to shift the number of solutions per board up,
+    while still making the end of the game unpredictable.
+    Note: this is *not* the same as the distribution of number of solutions.
+"""
+SOLN_DISTR = 0, 0.05, 0.05, 0.1, 0.15, 0.25, 0.2, 0.15, .05
+
+image_path = Path("bot", "resources", "evergreen", "all_cards.png")
+font_path = Path("bot", "resources", "evergreen", "LuckiestGuy-Regular.ttf")
+
+ALL_CARDS = Image.open(image_path)
+LABEL_FONT = ImageFont.truetype(str(font_path), size=16)
+CARD_WIDTH = 155
+CARD_HEIGHT = 97
+
+EMOJI_WRONG = "\u274C"
+
+ANSWER_REGEX = re.compile(r'^\D*(\d+)\D+(\d+)\D+(\d+)\D*$')
+
+
+def assemble_board_image(board: list[tuple[int]], rows: int, columns: int) -> Image:
+    """Cut and paste images representing the given cards into an image representing the board."""
+    new_im = Image.new("RGBA", (CARD_WIDTH*columns, CARD_HEIGHT*rows))
+    draw = ImageDraw.Draw(new_im)
+    for idx, card in enumerate(board):
+        card_image = get_card_image(card)
+        row, col = divmod(idx, columns)
+        top, left = row * CARD_HEIGHT, col * CARD_WIDTH
+        new_im.paste(card_image, (left, top))
+        draw.text(
+            xy=(left+5, top+5),  # magic numbers are buffers for the card labels
+            text=str(idx),
+            fill=(0, 0, 0),
+            font=LABEL_FONT,
+        )
+    return new_im
+
+
+def get_card_image(card: tuple[int]) -> Image:
+    """Slice the image containing all the cards to get just this card."""
+    row, col = divmod(as_trinary(card), 9)  # all_cards.png should have 9x9 cards
+    x1 = col * CARD_WIDTH
+    x2 = x1 + CARD_WIDTH
+    y1 = row * CARD_HEIGHT
+    y2 = y1 + CARD_HEIGHT
+    return ALL_CARDS.crop((x1, y1, x2, y2))
+
+
+def as_trinary(card: tuple[int]) -> int:
+    """Find the card's unique index by interpreting its features as trinary."""
+    return int(''.join(str(x) for x in card), base=3)
+
+
+class DuckGame:
+    """A class for a single game."""
+
+    def __init__(self,
+                 rows: int = 4,
+                 columns: int = 3,
+                 minimum_solutions: int = 1,
+                 ) -> None:
```

Comment:
I would usually format something like this like
```python
    def __init__(
        self,
        rows: int = 4,
        columns: int = 3,
        minimum_solutions: int = 1,
    ) -> None:
```
this is what the style guide suggests too https://pythondiscord.com/pages/guides/pydis-guides/contributing/style-guide/#bracket-and-item-arrangement